updater/src/updater.rs

```diff
@@ -161,41 +184,68 @@ impl Updater {
+			return Err(format!("Current executable ({}) is unreleased.", self.this.hash));
+		}
 
-			let latest_in_track = operations.latest_in_track(CLIENT_ID, self.track().into())?;
-			let in_track = Self::collect_release_info(operations, &latest_in_track)?;
-			let mut in_minor = Some(in_track.clone());
-			const PROOF: &'static str = "in_minor initialised and assigned with Some; loop breaks if None assigned; qed";
-			while in_minor.as_ref().expect(PROOF).version.track != self.track() {
-				let track = match in_minor.as_ref().expect(PROOF).version.track {
-					ReleaseTrack::Beta => ReleaseTrack::Stable,
-					ReleaseTrack::Nightly => ReleaseTrack::Beta,
-					_ => { in_minor = None; break; }
-				};
-				in_minor = Some(Self::collect_release_info(operations, &operations.latest_in_track(CLIENT_ID, track.into())?)?);
-			}
+		let client = self.client.upgrade().ok_or("Cannot obtain client")?;
+		let address = client.registry_address("operations".into()).ok_or("Cannot get operations contract address")?;
+		let do_call = |data| client.call_contract(BlockId::Latest, address, data).map_err(|e| format!("{:?}", e));
+
+		trace!(target: "updater", "Looking up this_fork for our release: {}/{:?}", CLIENT_ID, self.this.hash);
+
+		// get the fork number of this release
+		let this_fork = self.operations_contract.functions()
+			.release()
+			.call(client_id_hash(), self.this.hash, &do_call)
+			.ok()
+			.and_then(|(fork, track, _, _)| {
+				let this_track: ReleaseTrack = (track.low_u64() as u8).into();
+				match this_track {
+					ReleaseTrack::Unknown => None,
+					_ => Some(fork.low_u64()),
+				}
+			});
+
+		// get the hash of the latest release in our track
+		let latest_in_track = self.operations_contract.functions()
+			.latest_in_track()
+			.call(client_id_hash(), u8::from(self.track()), &do_call)
+			.map_err(|e| format!("{:?}", e))?;
+
+
+		// get the release info for the latest version in track
+		let in_track = self.collect_release_info(latest_in_track)?;
+		let mut in_minor = Some(in_track.clone());
+		const PROOF: &'static str = "in_minor initialised and assigned with Some; loop breaks if None assigned; qed";
+
+		// if the minor version has changed, let's check the minor version on a different track
+		while in_minor.as_ref().expect(PROOF).version.version.minor != self.this.version.minor {
+			let track = match self.track() {
+				ReleaseTrack::Beta => ReleaseTrack::Stable,
+				ReleaseTrack::Nightly => ReleaseTrack::Beta,
+				_ => { in_minor = None; break; }
+			};
+
+			let latest_in_track = self.operations_contract.functions()
+				.latest_in_track()
+				.call(client_id_hash(), u8::from(track), &do_call)
+				.map_err(|e| format!("{:?}", e))?;
 
-			Ok(OperationsInfo {
-				fork: operations.latest_fork()? as u64,
-				this_fork: this_fork,
-				track: in_track,
-				minor: in_minor,
-			})
-		} else {
-			Err("Operations not available".into())
+			in_minor = Some(self.collect_release_info(latest_in_track)?);
```

Comment:
Imho would be cleaner to use `loop` instead of `while` (and enclose the whole thing in a function).
```rust
let (in_track, in_minor) = self.find_versions(self.track());

// impl
fn find_versions(&self, current_track: _) -> (_, Option<_>) {
  let mut version_in_track = None;
  let mut track = current_track;
  loop {
    let latest_in_track = ...;
    let in_minor = self.collect_release_info(latest_in_track)?;
    // update initial version_in_track
    version_in_track = version_in_track.or_else(|| Some(in_minor.clone()));
    // check version
    if in_minor.version.version.minor == self.this.version.minor {
       return (version_in_track.unwrap(), Some(in_minor));
    }
    // check next track
    track = match in_minor.track {
      ReleaseTrack::Beta => ReleaseTrack::Stable,
      ReleaseTrack::Nightly => ReleaseTrack::Beta,
      _ => return (version_in_track.unwrap(), None),
    };
  }
}

Reply:
One more idea:
```rust

let in_track = self.find_version(self.track());
let in_minor = self.find_minor_version(in_track);


// impl
fn find_minor_version(&self, start_track: _) -> Option<_> {
   let mut track = start_track;
   while track.version.version.minor != self.this.version.minor {
     let next_track = match track.track {
      ReleaseTrack::Beta => ReleaseTrack::Stable,
      ReleaseTrack::Nightly => ReleaseTrack::Beta,
      _ => return None,
     };
     track = self.find_version(next_track);
   }
   return Some(track);
}

fn find_version(track: _) -> _ {
    let release_info = ...;
    self.collect_release_info(release_info)?;
}